File: crates/tanstack-ai/tests/test_utils.rs
//! Test utilities for tanstack-ai tests.
//!
//! Provides a mock adapter, chunk factories, and helpers mirroring
//! the TypeScript test-utils.ts.

use async_trait::async_trait;
use futures_util::stream;
use std::sync::{Arc, Mutex};
use tanstack_ai::adapter::{ChunkStream, TextAdapter};
use tanstack_ai::error::AiResult;
use tanstack_ai::types::*;

// ============================================================================
// Chunk factory helpers
// ============================================================================

pub fn now() -> f64 {
    chrono::Utc::now().timestamp_millis() as f64 / 1000.0
}

pub fn run_started(run_id: &str) -> StreamChunk {
    StreamChunk::RunStarted {
        timestamp: now(),
        run_id: run_id.to_string(),
        thread_id: None,
        model: Some("mock-model".to_string()),
    }
}

pub fn text_start(message_id: &str) -> StreamChunk {
    StreamChunk::TextMessageStart {
        timestamp: now(),
        message_id: message_id.to_string(),
        role: "assistant".to_string(),
        model: Some("mock-model".to_string()),
    }
}

pub fn text_content(delta: &str, message_id: &str) -> StreamChunk {
    StreamChunk::TextMessageContent {
        timestamp: now(),
        message_id: message_id.to_string(),
        delta: delta.to_string(),
        content: None,
        model: Some("mock-model".to_string()),
    }
}

pub fn _text_content_with_full(delta: &str, full: &str, message_id: &str) -> StreamChunk {
    StreamChunk::TextMessageContent {
        timestamp: now(),
        message_id: message_id.to_string(),
        delta: delta.to_string(),
        content: Some(full.to_string()),
        model: Some("mock-model".to_string()),
    }
}

pub fn text_end(message_id: &str) -> StreamChunk {
    StreamChunk::TextMessageEnd {
        timestamp: now(),
        message_id: message_id.to_string(),
        model: Some("mock-model".to_string()),
    }
}

pub fn tool_start(tool_call_id: &str, tool_name: &str, index: Option<usize>) -> StreamChunk {
    StreamChunk::ToolCallStart {
        timestamp: now(),
        tool_call_id: tool_call_id.to_string(),
        tool_name: tool_name.to_string(),
        parent_message_id: None,
        index,
        provider_metadata: None,
        model: Some("mock-model".to_string()),
    }
}

pub fn tool_args(tool_call_id: &str, delta: &str) -> StreamChunk {
    StreamChunk::ToolCallArgs {
        timestamp: now(),
        tool_call_id: tool_call_id.to_string(),
        delta: delta.to_string(),
        args: None,
        model: Some("mock-model".to_string()),
    }
}

pub fn _tool_end(
    tool_call_id: &str,
    tool_name: &str,
    input: Option<serde_json::Value>,
    result: Option<String>,
) -> StreamChunk {
    StreamChunk::ToolCallEnd {
        timestamp: now(),
        tool_call_id: tool_call_id.to_string(),
        tool_name: tool_name.to_string(),
        input,
        result,
        model: Some("mock-model".to_string()),
    }
}

pub fn run_finished(finish_reason: &str, run_id: &str) -> StreamChunk {
    StreamChunk::RunFinished {
        timestamp: now(),
        run_id: run_id.to_string(),
        finish_reason: Some(finish_reason.to_string()),
        usage: None,
        model: Some("mock-model".to_string()),
    }
}

pub fn run_error(message: &str, run_id: &str) -> StreamChunk {
    StreamChunk::RunError {
        timestamp: now(),
        run_id: Some(run_id.to_string()),
        error: RunErrorData {
            message: message.to_string(),
            code: None,
        },
        model: Some("mock-model".to_string()),
    }
}

pub fn step_finished(delta: &str, step_id: &str) -> StreamChunk {
    StreamChunk::StepFinished {
        timestamp: now(),
        step_id: step_id.to_string(),
        delta: delta.to_string(),
        content: None,
        model: Some("mock-model".to_string()),
    }
}

// ============================================================================
// Mock adapter
// ============================================================================

/// Tracks calls made to the mock adapter.
#[derive(Debug, Clone)]
pub struct MockCall {
    pub messages: Vec<ModelMessage>,
    pub _tools: Vec<String>,
    pub system_prompts: Vec<String>,
    pub temperature: Option<f64>,
    pub top_p: Option<f64>,
    pub max_tokens: Option<u32>,
}

/// A mock adapter that returns predetermined chunks per iteration.
pub struct MockAdapter {
    iterations: Vec<Vec<StreamChunk>>,
    calls: Arc<Mutex<Vec<MockCall>>>,
    call_index: Arc<Mutex<usize>>,
}

impl MockAdapter {
    pub fn new(iterations: Vec<Vec<StreamChunk>>) -> Self {
        Self {
            iterations,
            calls: Arc::new(Mutex::new(Vec::new())),
            call_index: Arc::new(Mutex::new(0)),
        }
    }

    pub fn calls(&self) -> Vec<MockCall> {
        self.calls.lock().unwrap().clone()
    }

    pub fn call_count(&self) -> usize {
        self.calls.lock().unwrap().len()
    }
}

#[async_trait]
impl TextAdapter for MockAdapter {
    fn name(&self) -> &str {
        "mock"
    }

    fn model(&self) -> &str {
        "mock-model"
    }

    async fn chat_stream(&self, options: &TextOptions) -> AiResult<ChunkStream> {
        // Record the call
        let call = MockCall {
            messages: options.messages.clone(),
            _tools: options.tools.iter().map(|t| t.name.clone()).collect(),
            system_prompts: options.system_prompts.clone(),
            temperature: options.temperature,
            top_p: options.top_p,
            max_tokens: options.max_tokens,
        };
        self.calls.lock().unwrap().push(call);

        // Get chunks for this iteration
        let mut idx = self.call_index.lock().unwrap();
        let chunks = self.iterations.get(*idx).cloned().unwrap_or_default();
        *idx += 1;

        let stream = stream::iter(chunks.into_iter().map(Ok));
        Ok(Box::pin(stream))
    }

    async fn structured_output(
        &self,
        _options: &StructuredOutputOptions,
    ) -> AiResult<StructuredOutputResult> {
        Ok(StructuredOutputResult {
            data: serde_json::json!({}),
            raw_text: "{}".to_string(),
        })
    }
}

// ============================================================================
// Collect helper
// ============================================================================

/// Collect all text content from chunks.
pub fn collect_text(chunks: &[StreamChunk]) -> String {
    let mut content = String::new();
    for chunk in chunks {
        if let StreamChunk::TextMessageContent {
            delta,
            content: full,
            ..
        } = chunk
        {
            if let Some(full_content) = full {
                content = full_content.clone();
            } else {
                content.push_str(delta);
            }
        }
    }
    content
}

/// Count chunks of a specific type.
pub fn count_chunk_type(chunks: &[StreamChunk], chunk_type: &str) -> usize {
    chunks
        .iter()
        .filter(|c| match chunk_type {
            "RUN_STARTED" => matches!(c, StreamChunk::RunStarted { .. }),
            "RUN_FINISHED" => matches!(c, StreamChunk::RunFinished { .. }),
            "TEXT_MESSAGE_START" => matches!(c, StreamChunk::TextMessageStart { .. }),
            "TEXT_MESSAGE_CONTENT" => matches!(c, StreamChunk::TextMessageContent { .. }),
            "TEXT_MESSAGE_END" => matches!(c, StreamChunk::TextMessageEnd { .. }),
            "TOOL_CALL_START" => matches!(c, StreamChunk::ToolCallStart { .. }),
            "TOOL_CALL_ARGS" => matches!(c, StreamChunk::ToolCallArgs { .. }),
            "TOOL_CALL_END" => matches!(c, StreamChunk::ToolCallEnd { .. }),
            _ => false,
        })
        .count()
}

/// Helper to create a simple server tool for testing.
pub fn server_tool(name: &str, result: serde_json::Value) -> Tool {
    let result_clone = result.clone();
    Tool::new(name, format!("Test tool: {}", name)).with_execute(
        move |_args: serde_json::Value, _ctx: tanstack_ai::types::ToolExecutionContext| {
            let r = result_clone.clone();
            async move { Ok(r) }
        },
    )
}

/// Helper to create a server tool that captures its arguments.
pub fn capturing_tool(name: &str) -> (Tool, Arc<Mutex<Vec<serde_json::Value>>>) {
    let captured: Arc<Mutex<Vec<serde_json::Value>>> = Arc::new(Mutex::new(Vec::new()));
    let captured_clone = captured.clone();
    let tool = Tool::new(name, format!("Capturing tool: {}", name)).with_execute(
        move |args: serde_json::Value, _ctx: tanstack_ai::types::ToolExecutionContext| {
            let c = captured_clone.clone();
            async move {
                c.lock().unwrap().push(args.clone());
                Ok(args)
            }
        },
    );
    (tool, captured)
}

File: packages/typescript/ai/tests/test-utils.ts
import { EventType } from '../src/types'
import type { AnyTextAdapter } from '../src/activities/chat/adapter'
import type {
  RunErrorEvent,
  RunFinishedEvent,
  RunStartedEvent,
  StepFinishedEvent,
  StepStartedEvent,
  StreamChunk,
  TextMessageContentEvent,
  TextMessageEndEvent,
  TextMessageStartEvent,
  Tool,
  ToolCallArgsEvent,
  ToolCallEndEvent,
  ToolCallStartEvent,
} from '../src/types'

// ============================================================================
// Chunk factory
// ============================================================================

/** Builds a typed StreamChunk by event type. Narrows the return to the
 *  matching variant via `Extract`, so callers get the right shape and TS
 *  catches missing required fields. Pass `EventType.X` for `type`. */
export function chunk<T extends StreamChunk['type']>(
  type: T,
  fields?: Record<string, unknown>,
): Extract<StreamChunk, { type: T }> {
  return { type, timestamp: Date.now(), ...fields } as Extract<
    StreamChunk,
    { type: T }
  >
}

// ============================================================================
// Event shorthand builders
// ============================================================================

/** Shorthand chunk factories for common AG-UI events. */
export const ev = {
  runStarted: (runId = 'run-1', threadId = 'thread-1'): RunStartedEvent => ({
    type: EventType.RUN_STARTED,
    runId,
    threadId,
    timestamp: Date.now(),
  }),
  textStart: (messageId = 'msg-1'): TextMessageStartEvent => ({
    type: EventType.TEXT_MESSAGE_START,
    messageId,
    role: 'assistant',
    timestamp: Date.now(),
  }),
  textContent: (
    delta: string,
    messageId = 'msg-1',
  ): TextMessageContentEvent => ({
    type: EventType.TEXT_MESSAGE_CONTENT,
    messageId,
    delta,
    timestamp: Date.now(),
  }),
  textEnd: (messageId = 'msg-1'): TextMessageEndEvent => ({
    type: EventType.TEXT_MESSAGE_END,
    messageId,
    timestamp: Date.now(),
  }),
  toolStart: (
    toolCallId: string,
    toolCallName: string,
    index?: number,
  ): ToolCallStartEvent => ({
    type: EventType.TOOL_CALL_START,
    toolCallId,
    toolCallName,
    toolName: toolCallName,
    timestamp: Date.now(),
    ...(index !== undefined ? { index } : {}),
  }),
  toolArgs: (toolCallId: string, delta: string): ToolCallArgsEvent => ({
    type: EventType.TOOL_CALL_ARGS,
    toolCallId,
    delta,
    timestamp: Date.now(),
  }),
  toolEnd: (
    toolCallId: string,
    toolCallName: string,
    opts?: { input?: unknown; result?: string },
  ): ToolCallEndEvent => ({
    type: EventType.TOOL_CALL_END,
    toolCallId,
    toolCallName,
    toolName: toolCallName,
    timestamp: Date.now(),
    ...opts,
  }),
  runFinished: (
    finishReason:
      | 'stop'
      | 'length'
      | 'content_filter'
      | 'tool_calls'
      | null = 'stop',
    runId = 'run-1',
    usage?: {
      promptTokens: number
      completionTokens: number
      totalTokens: number
    },
    threadId = 'thread-1',
  ): RunFinishedEvent => ({
    type: EventType.RUN_FINISHED,
    runId,
    threadId,
    finishReason,
    timestamp: Date.now(),
    ...(usage ? { usage } : {}),
  }),
  runError: (message: string): RunErrorEvent => ({
    type: EventType.RUN_ERROR,
    message,
    timestamp: Date.now(),
    error: { message },
  }),
  stepStarted: (stepName = 'step-1'): StepStartedEvent => ({
    type: EventType.STEP_STARTED,
    stepName,
    timestamp: Date.now(),
  }),
  stepFinished: (delta: string, stepName = 'step-1'): StepFinishedEvent => ({
    type: EventType.STEP_FINISHED,
    stepName,
    stepId: stepName,
    delta,
    timestamp: Date.now(),
  }),
}

// ============================================================================
// Mock adapter
// ============================================================================

/**
 * Create a mock adapter that satisfies AnyTextAdapter.
 * `chatStreamFn` receives the options and returns an AsyncIterable of chunks.
 * Multiple invocations can be tracked via the returned `calls` array.
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any -- mock adapter callbacks receive internal SDK types
export function createMockAdapter(options: {
  chatStreamFn?: (opts: any) => AsyncIterable<StreamChunk>
  /** Array of chunk sequences: chatStream returns iterations[0] on first call, iterations[1] on second, etc. */
  iterations?: Array<Array<StreamChunk>>
  structuredOutput?: (opts: any) => Promise<{ data: unknown; rawText: string }>
  /**
   * Override `adapter.chat()` for non-streaming tests. If omitted the
   * BaseTextAdapter default (drains chatStream) applies — which is fine
   * for tests that only care that *some* chat() impl runs.
   */
  chatFn?: (opts: any) => Promise<{
    content: string
    reasoning?: string
    toolCalls?: Array<{
      id: string
      type: 'function'
      function: { name: string; arguments: string }
    }>
    finishReason?: string
    usage?: {
      promptTokens?: number
      completionTokens?: number
      totalTokens?: number
    }
  }>
  /** Array of chat() responses: chatFn returns chatIterations[0] on first call, etc. */
  chatIterations?: Array<{
    content: string
    reasoning?: string
    toolCalls?: Array<{
      id: string
      type: 'function'
      function: { name: string; arguments: string }
    }>
    finishReason?: string
    usage?: {
      promptTokens?: number
      completionTokens?: number
      totalTokens?: number
    }
  }>
}) {
  const calls: Array<Record<string, unknown>> = []
  const chatCalls: Array<Record<string, unknown>> = []
  let callIndex = 0
  let chatCallIndex = 0

  const adapter: AnyTextAdapter = {
    kind: 'text' as const,
    name: 'mock',
    model: 'test-model' as const,
    '~types': {
      providerOptions: {} as Record<string, unknown>,
      inputModalities: ['text'] as readonly ['text'],
      messageMetadataByModality: {
        text: undefined as unknown,
        image: undefined as unknown,
        audio: undefined as unknown,
        video: undefined as unknown,
        document: undefined as unknown,
      },
      toolCapabilities: [] as ReadonlyArray<string>,
      toolCallMetadata: undefined as unknown,
    },
    chatStream: (opts: any) => {
      calls.push(opts)

      if (options.chatStreamFn) {
        return options.chatStreamFn(opts)
      }

      if (options.iterations) {
        const chunks = options.iterations[callIndex] || []
        callIndex++
        return (async function* () {
          for (const c of chunks) yield c
        })()
      }

      return (async function* () {})()
    },
    chatNonStreaming: async (opts: any) => {
      chatCalls.push(opts)

      if (options.chatFn) {
        return options.chatFn(opts)
      }

      if (options.chatIterations) {
        const result = options.chatIterations[chatCallIndex] || {
          content: '',
        }
        chatCallIndex++
        return result
      }

      // Default: drain the mock's chatStream and reassemble — mirrors the
      // legacy stream-then-concatenate behaviour so existing tests that
      // configure `iterations` (stream chunks) keep working when invoked
      // through the non-streaming path.
      let content = ''
      let reasoning = ''
      const toolCallsByIndex = new Map<
        number,
        { id: string; name: string; args: string }
      >()
      let nextIndex = 0
      let finishReason: string | undefined
      let usage:
        | {
            promptTokens?: number
            completionTokens?: number
            totalTokens?: number
          }
        | undefined

      for await (const chunk of adapter.chatStream(opts)) {
        switch (chunk.type) {
          case 'TEXT_MESSAGE_CONTENT': {
            const e = chunk as { delta?: string }
            if (e.delta) content += e.delta
            break
          }
          case 'REASONING_MESSAGE_CONTENT': {
            const e = chunk as { delta?: string }
            if (e.delta) reasoning += e.delta
            break
          }
          case 'TOOL_CALL_START': {
            const e = chunk as {
              toolCallId: string
              toolCallName?: string
              toolName?: string
              index?: number
            }
            const index = e.index ?? nextIndex++
            toolCallsByIndex.set(index, {
              id: e.toolCallId,
              name: e.toolCallName ?? e.toolName ?? '',
              args: '',
            })
            break
          }
          case 'TOOL_CALL_ARGS': {
            const e = chunk as { toolCallId: string; delta?: string }
            for (const tc of toolCallsByIndex.values()) {
              if (tc.id === e.toolCallId) {
                tc.args += e.delta ?? ''
                break
              }
            }
            break
          }
          case 'RUN_FINISHED': {
            const e = chunk as {
              finishReason?: string
              usage?: typeof usage
            }
            if (e.finishReason) finishReason = e.finishReason
            if (e.usage) usage = e.usage
            break
          }
        }
      }

      const toolCalls =
        toolCallsByIndex.size > 0
          ? Array.from(toolCallsByIndex.values()).map((tc) => ({
              id: tc.id,
              type: 'function' as const,
              function: { name: tc.name, arguments: tc.args },
            }))
          : undefined

      return {
        content,
        ...(reasoning ? { reasoning } : {}),
        ...(toolCalls ? { toolCalls } : {}),
        ...(finishReason ? { finishReason } : {}),
        ...(usage ? { usage } : {}),
      }
    },
    structuredOutput:
      options.structuredOutput ?? (async () => ({ data: {}, rawText: '{}' })),
  }

  return { adapter, calls, chatCalls }
}

// ============================================================================
// Stream collection
// ============================================================================

/** Collect all chunks from an async iterable. */
export async function collectChunks(
  stream: AsyncIterable<StreamChunk>,
): Promise<Array<StreamChunk>> {
  const chunks: Array<StreamChunk> = []
  for await (const c of stream) {
    chunks.push(c)
  }
  return chunks
}

// ============================================================================
// Type guards & extraction helpers
// ============================================================================

/** Type guard for TEXT_MESSAGE_CONTENT chunks. */
export function isTextContent(c: StreamChunk): c is TextMessageContentEvent {
  return c.type === 'TEXT_MESSAGE_CONTENT'
}

/** Extract all text deltas from a chunk array. */
export function getDeltas(chunks: Array<StreamChunk>): Array<string> {
  return chunks.filter(isTextContent).map((c) => c.delta)
}

// ============================================================================
// Tool helpers
// ============================================================================

/** Simple server tool for testing. */
export function serverTool(
  name: string,
  executeFn: (args: unknown) => unknown,
): Tool {
  return {
    name,
    description: `Test tool: ${name}`,
    execute: executeFn,
  }
}

/** Client tool (no execute function). */
export function clientTool(
  name: string,
  opts?: { needsApproval?: boolean },
): Tool {
  return {
    name,
    description: `Client tool: ${name}`,
    needsApproval: opts?.needsApproval,
  }
}
